### ISSUE TITLE
Endpoints unreachable k8s 1.19.0

### ISSUE THREAD

**Issue by alex1989hu** [cilium/cilium#13096]
## Bug report

We wanted to upgrade our Kubernetes cluster from `v1.18.8` to `v1.19.0`. Updated the following order: API Server -> Controller Manager -> Scheduler. The last one is the `kubelet`. After updated the `kubelet` to `v1.19.0` on a worker node *(prod-k8s-worker-8da7da49)*, than readinessProbe fails, or the Pod can not reach `world`.

Downgrade worker nodes brings back a healthy state. Only kept one with never version for debug purpose. 

Let's check what I can see from the Pod running on new `kubelet`:

```bash
kubectl -n kube-system exec -ti cilium-hqr9x -- cilium status
KVStore:                     Ok   Disabled
Kubernetes:                  Ok   1.19 (v1.19.0) [linux/amd64]
Kubernetes APIs:             ["CustomResourceDefinition", "cilium/v2::CiliumClusterwideNetworkPolicy", "cilium/v2::CiliumEndpoint", "cilium/v2::CiliumNetworkPolicy", "cilium/v2::CiliumNode", "core/v1::Namespace", "core/v1::Node", "core/v1::Pods", "core/v1::Service", "discovery/v1beta1::EndpointSlice", "networking.k8s.io/v1::NetworkPolicy"]
KubeProxyReplacement:        Probe   [eth0 (DR)]   [NodePort (SNAT, 30000-32767, XDP: DISABLED), HostPort, ExternalIPs, HostReachableServices (TCP, UDP), SessionAffinity]
Cilium:                      Ok      OK
NodeMonitor:                 Listening for events on 8 CPUs with 64x4096 of shared memory
Cilium health daemon:        Ok
IPAM:                        IPv4: 7/255 allocated from 10.0.1.0/24,
Masquerading:                BPF   [eth0]   10.0.1.0/24
Controller Status:           39/39 healthy
Proxy Status:                OK, ip 10.0.1.177, 0 redirects active on ports 10000-20000
Hubble:                      Ok              Current/Max Flows: 3218/4096 (78.56%), Flows/s: 6.50   Metrics: Ok
Cluster health:              3/8 reachable   (2020-09-04T15:32:29Z)
  Name                       IP              Node        Endpoints
  prod-k8s-master-1          192.249.66.180    reachable   unreachable
  prod-k8s-master-2          192.249.66.181    reachable   unreachable
  prod-k8s-worker-8ddbacc3   192.249.66.188    reachable   unreachable
  prod-k8s-worker-f08f772d   192.249.66.187    reachable   unreachable
  prod-k8s-worker-fc988125   192.249.66.185    reachable   unreachable
```

Let's check what I can see from a Pod running on old `kubelet`:
```bash
$ kubectl -n kube-system exec -ti cilium-qtsf7 -- cilium status
KVStore:                     Ok   Disabled
Kubernetes:                  Ok   1.19 (v1.19.0) [linux/amd64]
Kubernetes APIs:             ["CustomResourceDefinition", "cilium/v2::CiliumClusterwideNetworkPolicy", "cilium/v2::CiliumEndpoint", "cilium/v2::CiliumNetworkPolicy", "cilium/v2::CiliumNode", "core/v1::Namespace", "core/v1::Node", "core/v1::Pods", "core/v1::Service", "discovery/v1beta1::EndpointSlice", "networking.k8s.io/v1::NetworkPolicy"]
KubeProxyReplacement:        Probe   [eth0 (DR)]   [NodePort (SNAT, 30000-32767, XDP: DISABLED), HostPort, ExternalIPs, HostReachableServices (TCP, UDP), SessionAffinity]
Cilium:                      Ok      OK
NodeMonitor:                 Listening for events on 8 CPUs with 64x4096 of shared memory
Cilium health daemon:        Ok
IPAM:                        IPv4: 19/255 allocated from 10.0.4.0/24,
Masquerading:                BPF   [eth0]   10.0.4.0/24
Controller Status:           102/102 healthy
Proxy Status:                OK, ip 10.0.4.117, 0 redirects active on ports 10000-20000
Hubble:                      Ok              Current/Max Flows: 4096/4096 (100.00%), Flows/s: 55.16   Metrics: Ok
Cluster health:              7/8 reachable   (2020-09-04T15:41:31Z)
  Name                       IP              Node        Endpoints
  prod-k8s-worker-8da7da49   192.249.66.186    reachable   unreachable
```

**General Information**

- Cilium version (run `cilium version`)
```
Client: 1.8.3 54cf3810d 2020-09-04T14:01:53+02:00 go version go1.14.7 linux/amd64
Daemon: 1.8.3 54cf3810d 2020-09-04T14:01:53+02:00 go version go1.14.7 linux/amd64
```
- Kernel version (run `uname -a`)
```
Linux prod-k8s-worker-8da7da49 5.8.5-talos #1 SMP Tue Sep 1 19:36:01 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```
- Orchestration system version in use (e.g. `kubectl version`, Mesos, ...)
```
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.6", GitCommit:"dff82dc0de47299ab66c83c626e08b245ab19037", GitTreeState:"clean", BuildDate:"2020-07-15T16:58:53Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.0", GitCommit:"e19964183377d0ec2052d1f1fa930c4d7575bd50", GitTreeState:"clean", BuildDate:"2020-08-26T14:23:04Z", GoVersion:"go1.15", Compiler:"gc", Platform:"linux/amd64"}
```
- Link to relevant artifacts (policies, deployments scripts, ...)
N/A
[cilium-sysdump-20200904-173209.zip](https://github.com/cilium/cilium/files/5175792/cilium-sysdump-20200904-173209.zip)

- Generate and upload a system zip:
```
curl -sLO https://git.io/cilium-sysdump-latest.zip && python cilium-sysdump-latest.zip
```

**How to reproduce the issue**
This is tricky because our playground does not produce this symptom - I believe because it does not have the same load, pods, network policies etc.

**Comment by aanm** [comment]
@alex1989hu can you provide the kubelet logs of the node that is failing? Thanks

**Comment by alex1989hu** [comment]
The updated node has been removed from the cluster to see what is happening with a freshly created worker. Recreated the worker node which brings the default `kubelet` v1.19.0. It tried to join but it failed *(and then the node is rebooted, it is Talos behaviour, ignore it).*

Tried 2x times:
- [add-new-worker-try-1.txt](https://github.com/cilium/cilium/files/5182344/add-new-worker-try-1.txt)
- [add-new-worker-try-2.txt](https://github.com/cilium/cilium/files/5182345/add-new-worker-try-2.txt)

**Comment by aanm** [comment]
@alex1989hu it would be helpful to have steps to replicate the issue as currently the GH issue does not provide any. I've set up a minikube cluster with Cilium 1.8 and I didn't see any issues in the cluster.

**Comment by aanm** [comment]
The only suspicious logs I can see in kubelet are:
```
reflector.go:424] k8s.io/client-go/informers/factory.go:134: watch of *v1.Service ended with: very short watch: k8s.io/client-go/informers/factory.go:134: Unexpected watch close - watch lasted less than a second and no items received
```

**Comment by alex1989hu** [comment]
I am constantly try to reproduce the issue in a new cluster. I was able to do in 1x time only which setup is: 3 master - 1 worker node. I am still looking for what is the recipe to make it 100% reproducible.

Also upgraded Kubernetes from `v1.19.0` to `v1.19.1` - same faulty behaviour.

Anyway, the faulty worker node logs many times this *(replaced IP by ([redacted-external-IP]))*:
> level=error msg="insertNeighbor failed" HardwareAddr= Interface=eth0 LinkIndex=4 error=timeout ipAddr=([redacted-external-IP]) reason="insertNeighbor arping failed" subsys=linux-datapath

Could you clarify what does it mean, please?

Let's check what does the cilium Pod says on that faulty worker node:

```bash
 # cilium-health status
Error: Cannot get status: Get "http://%2Fvar%2Frun%2Fcilium%2Fhealth.sock/v1beta/status": dial unix /var/run/cilium/health.sock: connect: no such file or directory

# cilium status
KVStore:                Ok   Disabled
Kubernetes:             Ok   1.19 (v1.19.1) [linux/amd64]
Kubernetes APIs:        ["CustomResourceDefinition", "cilium/v2::CiliumClusterwideNetworkPolicy", "cilium/v2::CiliumEndpoint", "cilium/v2::CiliumNetworkPolicy", "cilium/v2::CiliumNode", "core/v1::Namespace", "core/v1::Node", "core/v1::Pods", "core/v1::Service", "discovery/v1beta1::EndpointSlice", "networking.k8s.io/v1::NetworkPolicy"]
KubeProxyReplacement:   Probe   [eth0 (DR)]   [NodePort (SNAT, 30000-32767, XDP: DISABLED), HostPort, ExternalIPs, HostReachableServices (TCP, UDP), SessionAffinity]
Cilium:                 Ok      OK
NodeMonitor:            Listening for events on 8 CPUs with 64x4096 of shared memory
IPAM:                   IPv4: 5/255 allocated from 10.0.9.0/24,
Masquerading:           BPF   [eth0]   10.0.9.0/24
Controller Status:      29/29 healthy
Proxy Status:           OK, ip 10.0.9.88, 0 redirects active on ports 10000-20000
Hubble:                 Ok   Current/Max Flows: 4096/4096 (100.00%), Flows/s: 17.12   Metrics: Ok
Cluster health:                                                                       Warning   cilium-health daemon unreachable
```

**Comment by smira** [comment]
We were able to reproduce this issue in Kubernetes 1.18.6 cluster with Cilium 1.8.2 and 1.8.3. (Exactly same scenario works with Flannel).

The steps were:

1. Install Kubernetes 1.18.6, we have 3 master + 1 worker node cluster in our tests.
2. Start upgrading kube-apiserver daemonset by bumping the image version and updating command (due to hyperkube- -> go-runner change in 1.19.0).
3. In the middle of kube-apiserver rolling restart, but seems always to be the case when 2 out of 3 instances of kube-apiserver were restarted with the new version, 3rd instance of kube-apiserver never comes up. Around the same time all the k8s services and cilium-operator start reporting following errors:

```
level=error msg=k8sError error="github.com/cilium/cilium/pkg/k8s/watchers/node.go:60: Failed to watch *v1.Node: Get \"https://10.96.0.1:443/api/v1/nodes?allowWatchBookmarks=true&fieldSelector=metadata.name%3Dmaster-2&resourceVersion=2341&timeoutSeconds=593&watch=true\": dial tcp 10.96.0.1:443: connect: connection refused" subsys=k8s
E0910 17:59:27.871228       1 reflector.go:383] k8s.io/apiserver/pkg/server/dynamiccertificates/configmap_cafile_content.go:206: Failed to watch *v1.ConfigMap: Get https://10.96.0.1:443/api/v1/namespaces/kube-system/configmaps?allowWatchBookmarks=true&fieldSelector=metadata.name%3Dextension-apiserver-authentication&resourceVersion=2336&timeout=5m43s&timeoutSeconds=343&watch=true: dial tcp 10.96.0.1:443: connect: connection refused
```

At the same time kube-apiserver instances (2 out of 3) are perfectly healthy, and `kubectl` still works just fine. But as kube-controller-manager and kube-scheduler stop working, no new pods can be scheduled in the cluster and the cluster is broken.

**Comment by smira** [comment]
Snapshot of `get pods -n kube-system`:

```
NAME                               READY   STATUS             RESTARTS   AGE     IP           NODE       NOMINATED NODE   READINESS GATES
cilium-22hql                       1/1     Running            0          11m     172.21.0.5   worker-1   <none>           <none>
cilium-59x6t                       1/1     Running            0          11m     172.21.0.2   master-1   <none>           <none>
cilium-5gm5s                       1/1     Running            0          11m     172.21.0.3   master-2   <none>           <none>
cilium-kvt6n                       1/1     Running            0          11m     172.21.0.4   master-3   <none>           <none>
cilium-operator-754dd76d85-87wqm   1/1     Running            3          11m     172.21.0.2   master-1   <none>           <none>
coredns-669d45d65b-2m64z           1/1     Running            0          11m     10.0.1.41    master-1   <none>           <none>
coredns-669d45d65b-74dc6           1/1     Running            0          11m     10.0.1.28    master-1   <none>           <none>
kube-apiserver-4qqp9               1/1     Running            0          3m2s    172.21.0.4   master-3   <none>           <none>
kube-apiserver-b422k               1/1     Running            0          2m57s   172.21.0.3   master-2   <none>           <none>
kube-controller-manager-22252      1/1     Running            0          10m     10.0.2.150   master-3   <none>           <none>
kube-controller-manager-d9l42      1/1     Running            0          11m     10.0.1.218   master-1   <none>           <none>
kube-controller-manager-xfnnf      0/1     CrashLoopBackOff   4          11m     10.0.0.205   master-2   <none>           <none>
kube-proxy-6xh7s                   1/1     Running            0          11m     172.21.0.3   master-2   <none>           <none>
kube-proxy-dpwlm                   1/1     Running            0          11m     172.21.0.2   master-1   <none>           <none>
kube-proxy-gj2fp                   1/1     Running            0          11m     172.21.0.5   worker-1   <none>           <none>
kube-proxy-qnhpk                   1/1     Running            0          11m     172.21.0.4   master-3   <none>           <none>
kube-scheduler-c2wpx               1/1     Running            0          11m     10.0.0.192   master-2   <none>           <none>
kube-scheduler-lbq4w               1/1     Running            0          10m     10.0.2.97    master-3   <none>           <none>
kube-scheduler-s5skk               1/1     Running            1          11m     10.0.1.55    master-1   <none>           <none>
pod-checkpointer-cbxsh             1/1     Running            0          9m17s   172.21.0.3   master-2   <none>           <none>
pod-checkpointer-cbxsh-master-2    1/1     Running            0          4m7s    172.21.0.3   master-2   <none>           <none>
pod-checkpointer-lw74b             1/1     Running            0          8m59s   172.21.0.2   master-1   <none>           <none>
pod-checkpointer-lw74b-master-1    1/1     Running            0          3m50s   172.21.0.2   master-1   <none>           <none>
pod-checkpointer-t55gf             1/1     Running            0          9m12s   172.21.0.4   master-3   <none>           <none>
pod-checkpointer-t55gf-master-3    1/1     Running            0          4m2s    172.21.0.4   master-3   <none>           <none>
```

**Comment by aanm** [comment]
@smira how do you install / deploy kubernetes?

**Comment by smira** [comment]
@aanm this is under Talos talos-systems/talos, which in turn is using bootkube.

I can reproduce and report the events.

**Comment by aanm** [comment]
@smira please do.

**Comment by smira** [comment]
@aanm I reran my test suite five times with Cilium 1.8.2 and 1.8.3, and I'm not able to reproduce the issue anymore.

My guess is that my upgrade automation had some issues, as I can't explain it any other way.

**Comment by aanm** [comment]
@smira thanks for taking a look!

**Comment by alex1989hu** [comment]
Created a new cluster, many `cilium` Pods:`cilium status` reports Endpoints `unreachable`.  Do you see any suspicious in the attached `sysdump`?
[cilium-sysdump-20200926-093322.zip](https://github.com/cilium/cilium/files/5286337/cilium-sysdump-20200926-093322.zip)

**Comment by aanm** [comment]
@alex1989hu I've taken a look at the sysdump and I couldn't find any errors that could explain this. I would double check if you have iptables correctly configured since the only change between environments is Kubernetes versions. If endpoints are unreachable double check the packets at the source and destination nodes.

> I am constantly try to reproduce the issue in a new cluster. I was able to do in 1x time only which setup is: 3 master - 1 worker node. I am still looking for what is the recipe to make it 100% reproducible.
> 
> Also upgraded Kubernetes from `v1.19.0` to `v1.19.1` - same faulty behaviour.
> 
> Anyway, the faulty worker node logs many times this _(replaced IP by ([redacted-external-IP]))_:
> 
> > level=error msg="insertNeighbor failed" HardwareAddr= Interface=eth0 LinkIndex=4 error=timeout ipAddr=([redacted-external-IP]) reason="insertNeighbor arping failed" subsys=linux-datapath
> 
> Could you clarify what does it mean, please?
> 

This PR https://github.com/cilium/cilium/pull/13482 might solve the issue you saw.

**Comment by alex1989hu** [comment]
Hello @aanm, thank you. I will check it, do you have ETA for releasing `1.8.5`?

**Comment by aanm** [comment]
> Hello @aanm, thank you. I will check it, do you have ETA for releasing `1.8.5`?

@alex1989hu there isn't an ETA for `1.8.5` at the moment.

**Comment by alex1989hu** [comment]
I made lots of investigation and I found thee root cause. Let me share the results

VMware `ESXi 7.0` and latter introduced a new revision of `VMXNET3` Network Adapter.

Using `6.5.0 build 9298722` with Kernel `v5.5.15` => OK
Using `6.5.0 build 9298722` with Kernel `v5.8.10` => OK
Using `7.0.1 build 16850804` with Kernel `v5.5.15` => OK
Using `7.0.1 build 16850804` with Kernel `v5.8.10 / v5.8.15` => **NOK**

Until a potential fix from VMware we have to use `E1000E` as Network Adapter.

I am pretty confident that is not related to Cilium.

Thank you for your help
